### PR TITLE
[tt-train] Enable fused q_rope, qkv_assemble, kv_down_split in mla

### PR DESCRIPTION
**Pull request by lgalasTT**
## Summary

Wire three MLA fusions into DeepSeek training in `mla.py`, replacing composite Python/TTNN paths with single autograd ops:

```python
# kv_down_split — split wkv_a output into kv latent + k_pe
kv, k_pe = autograd_split(kv_full, [kv_lora, qk_rope], dim=3)

# qkv_assemble — head-split Q, demux k_nope/v, broadcast k_pe into K
q, k_full, v = ttml.ops.mla.qkv_assemble(
    q_pre, kv_up, k_pe, n_heads, qk_nope, qk_rope, v_dim
)

# q_rope — RoPE on Q's rope suffix only
q_full = ttml.ops.rope.mla_q_rope(q, self.rope_params, qk_nope, qk_rope)
```

Previously these stages were implemented as many separate ops per forward/backward pass (slices, reshapes, concats, and a standalone RoPE). The fused paths cut kernel dispatch overhead and intermediate tensor traffic without changing model semantics.

## Motivation

Three hotspots were identified in the MLA forward path ([`mla_profile_benchmark.py`](https://github.com/tenstorrent/tt-metal/pull/45036)):

| Stage | Composite path | Fused replacement | 
|-------|----------------|-----------------|
| **kv_down_split** | Two `autograd_slice` on `wkv_a` output | `autograd_split` (single concat backward) |
| **qkv_assemble** | `split_heads` + slice + concat + `k_pe` broadcast | `ttml.ops.mla.qkv_assemble` |
| **q_rope** | `autograd_slice` + `rope` + `autograd_concat` | `ttml.ops.rope.mla_q_rope` |

## What changed

**File:** `tt-train/sources/ttml/ttml/models/deepseek/mla.py`

### kv_down_split

| Before | After |
|--------|-------|
| `autograd_slice` → `kv` | `autograd_split(kv_full, [kv_lora, qk_rope], dim=3)` |
| `autograd_slice` → `k_pe` | (same op; both outputs returned) |

Backward: two adjacent slices each rebuild a full-width grad (zeros + concat) that are then summed. `autograd_split` backward is a single `ttnn.concat` of chunk gradients.

### qkv_assemble

| Before | After |
|--------|-------|
| `split_heads(q_pre)` | |
| `split_heads(kv_up)` | `ttml.ops.mla.qkv_assemble(q_pre, kv_up, k_pe, …)` |
| Slice `k_nope`, slice `v` from `kv_up` | → `q`, `k_full`, `v` |
| `autograd_concat([k_pe] * n_heads, dim=1)` | |
| `autograd_concat([k_nope, k_pe_broadcast], dim=3)` | |

Pure data movement: head-split Q (not yet RoPE'd), demux per-head `k_nope`/`v`, broadcast shared `k_pe` into each head's K rope suffix.

### q_rope

| Before | After |
|--------|-------|
| Slice `q_nope` and `q_pe` from head-split `q` | Head-split `q` passed directly to fused op |
| `ttml.ops.rope.rope(q_pe, …)` | — |
| `autograd_concat([q_nope, q_pe], dim=3)` | `ttml.ops.rope.mla_q_rope(q, rope_params, qk_nope, qk_rope)` |

**Unchanged:**
- Q projection (`wq` / `wq_a → norm → wq_b`)
- `wkv_a`, `k_pe` RoPE, `wkv_b` up-projection
- Fused causal SDPA and output projection

## Correctness

- **kv_down_split:** `test_deepseek_autograd_split.py` — `autograd_split` matches dual `autograd_slice` forward/backward.
- **qkv_assemble:** `mla_qkv_assemble_op_test.cpp` — fused output matches composite `split_heads` + slice + concat reference within BF16 tolerance; autograd parity included.
- **q_rope:** `mla_q_rope_op_test.cpp` (`MLA_QRopeShapes/*`) — fused output matches composite slice + `rotary_embedding_llama` + concat reference.

## Training results

Compared **composite baseline** (dual-slice kv split + `qkv_assemble` + `q_rope`) vs **all three fusions enabled** on Blackhole. Each run: 1000 training steps; step time = mean over 985 measured steps (15-step warmup); final loss = average of last 100 steps.

| Config | Composite step (ms) | Fused step (ms) | Speedup | Composite loss | Fused loss |
|--------|--------------------:|----------------:|--------:|---------------:|-----------:|
| Nano DeepSeek, seq=256 | 334.46 | 255.28 | **1.31×** | 1.124297 | 1.128125 |
| Tiny DeepSeek, seq=512 | 1921.86 | 1785.54 | **1.08×** | 2.036484 | 2.025078 |
| Tiny DeepSeek, seq=1024 | 1923.17 | 1821.05 | **1.06×** | 1.858125 | 1.869999 |
| Tiny DeepSeek, seq=2048 | 3958.00 | 3415.14 | **1.16×** | 1.280000 | 1.281016 |

Combined fusion yields **~6–31%** end-to-end step-time reduction vs the full composite MLA path, with matched loss convergence.

All configs has batch = 1 for tiny and 32 for nano if not stated otherwise.

#### Tiny DeepSeek, seq=2048, batch=4
**Step time** — composite 3958 ms, fused 3415 ms (1.16×)
step_time
**Loss** — composite 1.280, fused 1.281 (last 100 steps avg)
losses

#### Tiny DeepSeek, seq=1024, batch=1
**Step time** — composite 1923 ms, fused 1821 ms (1.06×)
step_time
**Loss** — composite 1.858, fused 1.870 (last 100 steps avg)
losses

#### Tiny DeepSeek, seq=512, batch=1
**Step time** — composite 1922 ms, fused 1786 ms (1.08×)
step_time
**Loss** — composite 2.036, fused 2.025 (last 100 steps avg)
losses

##### Nano DeepSeek, seq=256
**Step time** — composite 334 ms, fused 255 ms (1.31×)
step_time
**Loss** — composite 1.124, fused 1.128 (last 100 steps avg)
losses

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:lgalasTT/enable_mla_q_rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:lgalasTT/enable_mla_q_rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:lgalasTT/enable_mla_q_rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:lgalasTT/enable_mla_q_rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:lgalasTT/enable_mla_q_rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:lgalasTT/enable_mla_q_rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:lgalasTT/enable_mla_q_rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:lgalasTT/enable_mla_q_rope)

### CODE REVIEW

**Review by mdragulaTT — CHANGES_REQUESTED** [review]
Looks good, but I think we should not merge this as a separate MLA wiring PR.

Let’s do one model-level integration PR that substitutes all ready MLA pieces together: `qkv_assemble`, `mla_q_rope`, and efficient `kv_down_split`. I don't see a strong reason to split it into multiple PRs.

**Review by mdragulaTT — APPROVED** [review]
Looks good now. One small cleanup before merge: please remove stale composite-path leftovers from `mla.py`:
- unused imports: `autograd_slice`, `autograd_concat`, `split_heads`
- unused locals: `B`, `S`, `qk_head`
- update the file/class comments that still describe `split_heads` -> `slice` -> `rope` -> `concat

**Review by philei-tt — APPROVED** [review]
Thanks, LGTM
